File: Difficulty: Hard/Longest valid Parentheses/longest-valid-parentheses.cpp
//{ Driver Code Starts
// Initial Template for C++

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// User function Template for C++

class Solution {
  public:
    int maxLength(string& str) {
        // code here
        // 16.09.24 Potd
        int open=0, close=0, ans=0, n=str.size();

        for(int i=0; i<n; i++){
            if(str[i] == '(')   open++;
            else                close++;
            
            if(close > open){
                open = close = 0;
            } else if(open == close){
                ans = max(ans, close*2);
            }
        }
        
        open = close = 0;
        for(int i=n-1; i>=0; i--){
            if(str[i] == '(')   open++;
            else                close++;
            
            if(open > close){
                open = close = 0;
            } else if(open == close){
                ans = max(ans, close*2);
            }
        }
        
        return ans;
    }
};

//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {
        string S;
        cin >> S;

        Solution ob;
        cout << ob.maxLength(S) << "\n";
    }
    return 0;
}
// } Driver Code Ends

File: Difficulty: Hard/Longest valid Parentheses/longest-valid-parentheses.py

class Solution:
    def maxLength(self, s):
        # code here
        # 22.02.25 POtd
        ans = 0
        stk = [-1]
        
        for i in range(len(s)):
            if s[i] == '(':
                stk.append(i)
            elif stk:
                stk.pop()
                if stk:
                    ans = max(ans, i - stk[-1])
                else:
                    stk.append(i)
                    
        return ans
        

#{ 
 # Driver Code Starts
# Initial Template for Python3

if __name__ == '__main__':
    t = int(input())
    for _ in range(t):
        S = input()

        ob = Solution()
        print(ob.maxLength(S))
        print("~")

# } Driver Code Ends